### ISSUE TITLE
False positive when splatting keyword arguments

### ISSUE THREAD

**Issue by billziss-gh** [microsoft/pyright#11074]
**Describe the bug**

Pyright may produce what appears to be a false positive when splatting keyword arguments. For example, this happens when splatting keyword arguments into [`numpy.savez`](https://numpy.org/doc/stable/reference/generated/numpy.savez.html).

**Code or Screenshots**

Minimal code sample in [pyright playground](https://pyright-play.net/?reportUnreachable=true&code=CYUwZgBAzghgbiAXgCjASwDYgDQQFQwBOA5lLjBhgPYDuA%2BgA5oDGA1lgLwAqhArjvjysawKAEoAXACgIsiA0JoAdgBdUAIgDe6TgF91YmXIXK1YLUVId9hufMWqNmitXpM2eg0dknH5zcKi1l5SAMQQAIIkvAC2IKoQVJAqAJ4MIBDqpuoQzDBKSlQqEABGGTBQUGjESiDAECpU8kQwcSoghJkutIws7CA5SQ1pGeolVFQYOcoQYLxKzCpoVEqZsAiI6mF2mdkQaFAQhcUVVTUwJVgNTWMTU1LrSMjqYBPquHh4mmNE6hIQAEZdGIgA).

```python
def savez(file, *args, allow_pickle=True, **kwds):
    print(f"{file=}")
    print(f"{args=}")
    print(f"{allow_pickle=}")
    print(f"{kwds=}")

# Argument of type "int" cannot be assigned to parameter "allow_pickle" of type "bool" in function "savez"
#     "int" is not assignable to "bool"
savez("foo", **{"bar": 1})
```

**VS Code extension or command-line**

VSCode + Pylance

**Possibly related**

#5545

**Comment by erictraut** [comment]
Pyright is working as designed here.

The inferred type of the `allow_pickle` parameter is `bool`, and the inferred type of the arg expression `{"bar": 1}` is `dict[str, int]`. This results in a type violation, which pyright is reporting.

You can avoid the type error by not using splatting here and instead directly using a keyword argument `savez("foo", bar=1)`.

**Comment by billziss-gh** [comment]
@erictraut

I understand that pyright is working as designed, but in my opinion the behavior is wrong and surprising.

> The inferred type of the allow_pickle parameter is bool, and the inferred type of the arg expression {"bar": 1} is dict[str, int]. This results in a type violation, which pyright is reporting.

Pyright reports that the `allow_pickle` parameter is being assigned but that is not the case. If pyright is not able to correctly determine the parameters assigned because of the dynamic nature of the splat operation that is understandable, but in this case it should not report any error at all.

In any case thank you for your time.
